### PR TITLE
Use compare cache for dimensions in incremental index

### PR DESCRIPTION
**Pull request by navis**
NavigableMap in IncrementalIndex frequently compares values to sort input rows. If cardinality is not that high, we can cache the compare result and use it again.

### CODE REVIEW

**Comment by navis** [comment]
Marked WIP because need modification on DimensionsSpec to accept per dimension configuration. With 2K cache (About 2M for each dimension), got 5~10 % performance enhancement of on-heap ingestion.

BEFORE (msec)

6590 
4454 
4774 
4713 
4619 
4583 
4884 
4545 
4566 
4381 
4565 
4442 

AFTER (msec)
4502
4007
4461
4430
3909
4020
4211
4178
4224
4136
4381
4072

**Comment by navis** [comment]
added `compareCacheSize` in `StringDimensionSchema`

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 1]
```diff
@@ -1081,9 +1094,14 @@ public ColumnCapabilitiesImpl getCapabilities()
...
 
-    NullValueConverterDimDim(DimDim delegate)
+    private final int capacity;
+    private final byte[] cache;
+
+    NullValueConverterDimDim(DimDim delegate, int cacheCapacity)
     {
       this.delegate = delegate;
+      this.capacity = Math.min(cacheCapacity, 4096);  // max 2M
+      this.cache = cacheCapacity > 0 ? new byte[(cacheCapacity * (cacheCapacity - 1) + 8) / 8] : null;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: should the mem allocation be based on this.capacity instead of cacheCapacity ?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 2]
```diff
@@ -1129,10 +1147,65 @@ public int add(String value)
...
+      int encoded = value == POSITIVE ? ENCODED_POSITIVE : value == NEGATIVE ? ENCODED_NEGATIVE : 0;
+      if (encoded != 0) {
+        cache[cacheIndex[0]] |= (byte)(encoded << (cacheIndex[1] << 1));
+      }
+    }
+
+    private byte normalize(int compare, boolean leftToRight)
+    {
+      return leftToRight ^ compare > 0 ? NEGATIVE : POSITIVE;
+    }
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: can we add some unit tests for the cache ?

**Comment by nishantmonu51** [comment]
Looks good to me, would be good to have some more unit tests for the cache.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java [hunk 3]
```diff
@@ -1129,10 +1147,65 @@ public int add(String value)
...
+        0b00000011, 0b00001100, 0b00110000, 0b11000000
+    };
+
+    private static final byte POSITIVE = (byte) 1;
+    private static final byte NEGATIVE = (byte) -1;
+
+    private static final byte ENCODED_POSITIVE = 0x01;
+    private static final byte ENCODED_NEGATIVE = 0x02;
+
+    private int[] toIndex(int lhsIdx, int rhsIdx, boolean leftToRight)
```
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: can you write up some comments to explain how this cache works?

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 2 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: agree, this needs some thorough tests, maybe making it a separate class would make testing easier?

File: processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 4]
```diff
@@ -1976,7 +1976,10 @@ public void testGroupByWithAlphaNumericDimensionOrder()
         .setDimensions(
             Lists.<DimensionSpec>newArrayList(
                 new ExtractionDimensionSpec(
-                    "quality", "alias", new LookupExtractionFn(new MapLookupExtractor(map, false), false, null, false, false), null
+                    "quality",
```
Comment (xvrl) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: unless there are any code changes, I would rather keep reformatting for unrelated files out of this PR.

File: api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java [hunk 5]
```diff
@@ -27,15 +27,19 @@
...
+    return new StringDimensionSchema(name, -1);
   }
 
+  private final int compareCacheSize;
+
   @JsonCreator
   public StringDimensionSchema(
-      @JsonProperty("name") String name
+      @JsonProperty("name") String name,
+      @JsonProperty("compareCacheSize") int compareCacheSize
```
Comment (xvrl) [inline api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java]: I'm not sure if this should be part of the schema, it seems to be this is mainly a runtime optimization that we should be able to enable or disable independently.

**Comment by xvrl** [comment]
@navis can you think of edge cases where this cache may actually cause performance to be worse?

File: api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java (hunk 5 above)
Comment (xvrl) [inline api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java]: I wonder if this may also cause issues when merging segments where one segment may have a cache and the other one doesn't.

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: I've ran druid tests with hard coded cache size and passed all. agree on it's demanding serious test cases. I'll try

File: processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java (hunk 3 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java]: was easy to read when one byte for an entry. now it's using 2bit and things look cryptic. I'll add some comments.

File: processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 4 above)
Comment (navis) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: ok

File: api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java (hunk 5 above)
Comment (navis) [inline api/src/main/java/io/druid/data/input/impl/StringDimensionSchema.java]: I've also thought of that(where to configure?). but the effectivity of cache is based on the cardinality of the dimension and I've thought it should be configurable per dimension.

Is there a process directly merging two incremental index datum? If it's not like that, it'll not make a problem.

**Comment by navis** [comment]
@xvrl when the cardinality exceeds cache capacity, the effectiveness will start degrade. I'm thinking of disabling it when some threshold exceeds (3x of capacity?). But generally it should not affect performance in any time.
